src/main/java/io/openshift/booster/catalog/BoosterCatalogService.java

```diff
@@ -69,52 +70,58 @@
 
    private static final Logger logger = Logger.getLogger(BoosterCatalogService.class.getName());
 
-   private final ReentrantReadWriteLock reentrantLock = new ReentrantReadWriteLock();
-
-   private volatile List<Booster> boosters = Collections.emptyList();
+   private volatile Set<Booster> boosters = new ConcurrentSkipListSet<>(Comparator.comparing(Booster::getName));
 
    private final BoosterCatalogPathProvider provider;
+   
+   private final ExecutorService executor;
+   
+   private final CompletableFuture<Set<Booster>> result = new CompletableFuture<>();
+   private volatile boolean indexingStarted = false;
 
-   private BoosterCatalogService(BoosterCatalogPathProvider provider)
+   private BoosterCatalogService(BoosterCatalogPathProvider provider, ExecutorService executor)
    {
       this.provider = provider;
+      this.executor = executor;
    }
 
    /**
     * Indexes the existing YAML files provided by the {@link BoosterCatalogPathProvider} implementation
     */
-   public Path index()
+   public synchronized CompletableFuture<Set<Booster>> index()
    {
-      WriteLock lock = reentrantLock.writeLock();
-      Path catalogPath = null;
-      try
-      {
-         lock.lock();
-         catalogPath = provider.createCatalogPath();
-         this.boosters = indexBoosters(catalogPath);
+      if (!indexingStarted) {
+          indexingStarted = true;
+          CompletableFuture.runAsync(() -> {
```

Comment:
No, you can't use `this::doIndex` there (it doesn't satisfy the necessary `Supplier` interface) and lambdas don't handle exceptions.

So the only way to do that (according to StackOverflow) is to do it yourself.

And of course the `doIndex()` method could be removed but that would get rid of the log messages and I didn't want to change any of the existing functionality.